compiler/src/iree/compiler/Dialect/HAL/Transforms/ReplacePushConstants.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+namespace {
+
+// These must match what the runtime uses.
+#define IREE_PUSH_CONSTANT_EMULATION_SET 3
+#define IREE_PUSH_CONSTANT_EMULATION_BINDING 0
```

Comment:
Is this what you expect on the runtime side? I see `IREE_HAL_WEBGPU_PARAMS_BIND_GROUP_INDEX` for the `set`, but I'm not sure about the `binding`.

Reply:
yes - I'd use `IREE_HAL_WEBGPU_PARAMS_BIND_GROUP_INDEX` here (to match the same name in the runtime) so that it's easy to find one or the other, and we should add `IREE_HAL_WEBGPU_PARAMS_BINDING_INDEX` to 0

Reply:
I hesitated to use `WEBGPU` in this, since the code is not under a WebGPU specific directory (though maybe it should be)

Reply:
ah, good point, but this is going to be webgpu specific (or at least may have different implementations for different backends). I think it'd make sense to put the pass under Codegen/WGSL/ alongside SPIRV ones. The TargetBackends can use those passes directly and since the pass only runs on the executables when generating code it seems fine. Could also put it under Codegen/SPIRV/ for now.

Reply:
(in Metal, for example, we would be using argument buffers or something instead of uniform buffers, so similar technique but different implementation)

Reply:
Moved the pass to Codegen/WGSL/ and updated those names.